elastic/kibana
# [Response Ops][Alerting] Initial implementation of FAAD `AlertsClient` for writing generic AAD documents

Resolves https://github.com/elastic/kibana/issues/156442

## Summary

1. Adds `shouldWriteAlerts` flag to rule type registration which defaults to `false` if not set. This prevents duplicate AAD documents from being written for the rule registry rule types that had to register with the framework in order to get their resources installed on startup.
2. Initial implementation of `AlertsClient` which primarily functions as a proxy to the `LegacyAlertsClient`. It does 2 additional thing:
  a. When initialized with the active & recovered alerts from the previous execution (de-serialized from the task manager state), it queries the AAD index for the corresponding alert document.
  b. When returning the alerts to serialize into the task manager state, it builds the alert document and bulk upserts into the AAD index.

This PR does not opt any rule types into writing these generic docs but adds an example functional test  that does. To test it out with the ES query rule type, add the following

```
diff --git a/x-pack/plugins/stack_alerts/server/rule_types/es_query/rule_type.ts b/x-pack/plugins/stack_alerts/server/rule_types/es_query/rule_type.ts
index 214d2ee4b76..0439a576b03 100644
--- a/x-pack/plugins/stack_alerts/server/rule_types/es_query/rule_type.ts
+++ b/x-pack/plugins/stack_alerts/server/rule_types/es_query/rule_type.ts
@@ -187,5 +187,12 @@ export function getRuleType(
     },
     producer: STACK_ALERTS_FEATURE_ID,
     doesSetRecoveryContext: true,
+    alerts: {
+      context: 'stack',
+      shouldWrite: true,
+      mappings: {
+        fieldMap: {},
+      },
+    },
   };
 }
```

## To Verify
- Verify that rule registry rule types still work as expected
- Verify that non rule-registry rule types still work as expected
- Modify a rule type to register with FAAD and write alerts and verify that the alert documents look as expected.
